# Data Science Fundamentals Assessment

**YuvaIntern AI/ML Research Specialist Internship - Task 1**  
**Author:** [Your Name]  
**Repository:** `Data-Science-Fundamentals`  

---

## Overview
This notebook serves as a comprehensive foundational assessment covering core concepts in Data Science. It bridges theoretical knowledge, Python programming fundamentals, basic statistics, data structures, and best practices required for reproducible data analysis.

---

## Table of Contents
1. [Data Science Fundamentals & Lifecycle](#1-data-science-fundamentals--lifecycle)
2. [Python Programming Basics](#2-python-programming-basics)
3. [Statistical Foundations](#3-statistical-foundations)
4. [Data Types & Structures](#4-data-types--structures)
5. [Data Science Best Practices & Applied Analysis](#5-data-science-best-practices--applied-analysis)

## 1. Data Science Fundamentals & Lifecycle

### 1.1 The Data Science Lifecycle
The Data Science lifecycle is an iterative process used to extract insights and build actionable models from raw data.
1. **Problem Definition:** Understanding business objectives and defining measurable goals.
2. **Data Collection:** Gathering data from databases, APIs, web scraping, or flat files.
3. **Data Cleaning & Preprocessing:** Handling missing values, outliers, formatting errors, and inconsistencies.
4. **Exploratory Data Analysis (EDA):** Summarizing main characteristics, visualizing trends, and understanding underlying patterns.
5. **Data Modeling & Machine Learning:** Applying statistical models or ML algorithms to make predictions or classifications.
6. **Evaluation & Communication:** Validating model performance and interpreting findings for business stakeholders.
7. **Deployment & Maintenance:** Deploying models into production environments and monitoring performance over time.

---

### 1.2 Types of Data Analysis
- **Descriptive Analysis:** Focuses on *what happened* in the past using historical data summaries and visualizations.
- **Diagnostic Analysis:** Examines *why something happened* by evaluating cause-and-effect relationships and correlations.
- **Predictive Analysis:** Forecasts *what is likely to happen* using historical patterns, statistical models, and machine learning.
- **Prescriptive Analysis:** Recommends *what actions to take* to optimize business outcomes using optimization and decision algorithms.

---

### 1.3 Role of Data Science in Solving Real-World Problems
Data Science enables evidence-based decision-making across various industries:
- **Healthcare:** Disease prediction, personalized medicine, and medical imaging analysis.
- **Finance:** Fraud detection, algorithmic trading, and credit risk assessment.
- **E-Commerce & Retail:** Recommender systems, customer churn prediction, and dynamic pricing models.

In [4]:
Total_NO_Stu  =  70
Average_marks  =  75
Subject   ="Robotics and Aartificial intelligence"
Has_Stu_Comp = True

print("Data Types")
print("Total number of student :",type(Total_NO_Stu))
print("Average marks of student:",type(Average_marks ))
print("Subject :",type(Subject ))
print("Has student completed  the course:",type(Has_Stu_Comp))

val_a=int (input("Enter your marks"))
val_b=int(input("Enter your marks"))

sum_result =val_a + val_b
div_result =val_a / val_b
power_result  = val_a ** 2

print("\n--- Arithmetic Results ---")
print("Addition Result:", sum_result)
print("Division Result:", div_result)
print("Power Result:", power_result)

Data Types
Total number of student : <class 'int'>
Average marks of student: <class 'int'>
Subject : <class 'str'>
Has student completed  the course: <class 'bool'>

--- Arithmetic Results ---
Addition Result: 90
Division Result: 1.25
Power Result: 2500


In [5]:
def calculate_average(marks):
    if marks >=90:
        return "Grade A"
    elif marks >=80:
        return "Grade B"
    else:
        return "Grade C\n need to improve your marks"

student_marks = int(input("\nEnter your marks to calculate grade: "))
grade = calculate_average(student_marks)
print(f"Your grade is: {grade}")

ValueError: invalid literal for int() with base 10: ''

In [ ]:
# 1. Function declaration
def evaluate_student(marks):
    if marks >= 85:
        return "Grade A"
    elif marks >= 60:
        return "Grade B"
    else:
        return "Needs Improvement"

# 2. List of marks
marks_list = [90, 75, 50, 62, 40]

# 3. For loop
for mark in marks_list:
    result = evaluate_student(mark)
    print(f"Marks: {mark} -> Status: {result}")

Marks: 90 -> Status: Grade A
Marks: 75 -> Status: Grade B
Marks: 50 -> Status: Needs Improvement
Marks: 62 -> Status: Grade B
Marks: 40 -> Status: Needs Improvement


### 2.1 Functions, Conditional Logic, and Loops

In Data Science, we frequently use functions to modularize our code, conditional statements (`if-elif-else`) to filter data or apply business rules, and loops (`for`/`while`) to iterate over datasets.

**Objective:** Create a function that evaluates student marks and categorizes them into performance grades, then iterate through a dataset of student scores.

In [ ]:
# Function to evaluate performance based on marks
def check_student_status(marks):
    if marks >= 80:
        return "Distinction / Grade A"
    elif marks >= 60:
        return "First Class / Grade B"
    elif marks >= 40:
        return "Pass / Grade C"
    else:
        return "Fail / Needs Improvement"

# List of marks for multiple students
student_marks = [85, 92, 58, 42, 35, 78]

# Loop over the list to check each student's result
print("--- Student Evaluation Summary ---")
for marks in student_marks:
    result = check_student_status(marks)
    print(f"Marks: {marks} | Status: {result}")

--- Student Evaluation Summary ---
Marks: 85 | Status: Distinction / Grade A
Marks: 92 | Status: Distinction / Grade A
Marks: 58 | Status: Pass / Grade C
Marks: 42 | Status: Pass / Grade C
Marks: 35 | Status: Fail / Needs Improvement
Marks: 78 | Status: First Class / Grade B


### 2.2 Data Structures (Lists, Tuples, Dictionaries, Sets)

Data structures are used to store and organize collections of data effectively.
- **Lists (`[]`):** Ordered, mutable (modifiable), allows duplicate elements.
- **Tuples (`()`):** Ordered, immutable (unmodifiable), allows duplicate elements.
- **Dictionaries (`{key: value}`):** Key-value pairs, mutable, unordered/ordered by insertion.
- **Sets (`{}`):** Unordered, mutable, contains only unique elements.

In [ ]:
# 1. List
students = ["Amit", "Neha", "Pooja"]
students.append("Rohan")

# 2. Tuple
course_info = ("Data Science", "Task 1", 2026)

# 3. Dictionary
student_profile = {
    "name": "Amit",
    "score": 88,
    "status": "Passed"
}

# 4. Set (Removes duplicates)
raw_scores = {80, 90, 80, 75, 90}  # Output will be {80, 90, 75}

### 2.2 Data Structures (Lists, Tuples, Dictionaries, Sets)

Data structures are used to store and organize collections of data effectively.
- **Lists (`[]`):** Ordered, mutable (modifiable), allows duplicate elements.
- **Tuples (`()`):** Ordered, immutable (unmodifiable), allows duplicate elements.
- **Dictionaries (`{key: value}`):** Key-value pairs, mutable, unordered/ordered by insertion.
- **Sets (`{}`):** Unordered, mutable, contains only unique elements.

In [8]:
# 1. List (Mutable & Ordered)
students_list = ["Amit", "Neha", "Pooja"]
students_list.append("Rohan")  # Adding an element

# 2. Tuple (Immutable & Ordered)
course_tuple = ("Data Science", "Python", "2026")

# 3. Dictionary (Key-Value Pairs)
student_dict = {
    "name": "Amit",
    "subject": "AI/ML",
    "score": 88
}

# 4. Set (Unique Elements Only)
unique_scores_set = {80, 90, 80, 75, 90, 88}

# Output Print
print("List:", students_list)
print("Tuple:", course_tuple)
print("Dictionary:", student_dict)
print("Set (Duplicates Removed):", unique_scores_set)

List: ['Amit', 'Neha', 'Pooja', 'Rohan']
Tuple: ('Data Science', 'Python', '2026')
Dictionary: {'name': 'Amit', 'subject': 'AI/ML', 'score': 88}
Set (Duplicates Removed): {80, 88, 90, 75}


### 2.2 Data Structures (Lists, Tuples, Dictionaries, Sets)

Data structures are used to store and organize collections of data effectively.
- **Lists (`[]`):** Ordered, mutable (modifiable), allows duplicate elements.
- **Tuples (`()`):** Ordered, immutable (unmodifiable), allows duplicate elements.
- **Dictionaries (`{key: value}`):** Key-value pairs, mutable, key-indexed.
- **Sets (`{}`):** Unordered, mutable, contains only unique elements.

In [6]:
# 1. List (Mutable & Ordered)
students_list = ["Amit", "Neha", "Pooja"]
students_list.append("Rohan")

# 2. Tuple (Immutable & Ordered)
course_tuple = ("Data Science", "Python", "2026")

# 3. Dictionary (Key-Value Pairs)
student_dict = {
    "name": "Amit",
    "subject": "AI/ML",
    "score": 88
}

# 4. Set (Unique Elements Only)
unique_scores_set = {80, 90, 80, 75, 90, 88}

# Output Display
print("List:", students_list)
print("Tuple:", course_tuple)
print("Dictionary:", student_dict)
print("Set (Duplicates Removed):", unique_scores_set)

List: ['Amit', 'Neha', 'Pooja', 'Rohan']
Tuple: ('Data Science', 'Python', '2026')
Dictionary: {'name': 'Amit', 'subject': 'AI/ML', 'score': 88}
Set (Duplicates Removed): {80, 88, 90, 75}


In [7]:
# 1. List (Mutable & Ordered)
students_list = ["Amit", "Neha", "Pooja"]
students_list.append("Rohan")

# 2. Tuple (Immutable & Ordered)
course_tuple = ("Data Science", "Python", "2026")

# 3. Dictionary (Key-Value Pairs)
student_dict = {
    "name": "Amit",
    "subject": "AI/ML",
    "score": 88
}

# 4. Set (Unique Elements Only)
unique_scores_set = {80, 90, 80, 75, 90, 88}

# Output Display
print("List:", students_list)
print("Tuple:", course_tuple)
print("Dictionary:", student_dict)
print("Set (Duplicates Removed):", unique_scores_set)

List: ['Amit', 'Neha', 'Pooja', 'Rohan']
Tuple: ('Data Science', 'Python', '2026')
Dictionary: {'name': 'Amit', 'subject': 'AI/ML', 'score': 88}
Set (Duplicates Removed): {80, 88, 90, 75}


# 1. List (Mutable & Ordered)
students_list = ["Amit", "Neha", "Pooja"]
students_list.append("Rohan")

# 2. Tuple (Immutable & Ordered)
course_tuple = ("Data Science", "Python", "2026")

# 3. Dictionary (Key-Value Pairs)
student_dict = {
    "name": "Amit",
    "subject": "AI/ML",
    "score": 88
}

# 4. Set (Unique Elements Only)
unique_scores_set = {80, 90, 80, 75, 90, 88}

# Output Display
print("List:", students_list)
print("Tuple:", course_tuple)
print("Dictionary:", student_dict)
print("Set (Duplicates Removed):", unique_scores_set)

In [9]:
# 1. List (Mutable & Ordered)
students_list = ["Amit", "Neha", "Pooja"]
students_list.append("Rohan")

# 2. Tuple (Immutable & Ordered)
course_tuple = ("Data Science", "Python", "2026")

# 3. Dictionary (Key-Value Pairs)
student_dict = {
    "name": "Amit",
    "subject": "AI/ML",
    "score": 88
}

# 4. Set (Unique Elements Only)
unique_scores_set = {80, 90, 80, 75, 90, 88}

# Output Display
print("List:", students_list)
print("Tuple:", course_tuple)
print("Dictionary:", student_dict)
print("Set (Duplicates Removed):", unique_scores_set)

List: ['Amit', 'Neha', 'Pooja', 'Rohan']
Tuple: ('Data Science', 'Python', '2026')
Dictionary: {'name': 'Amit', 'subject': 'AI/ML', 'score': 88}
Set (Duplicates Removed): {80, 88, 90, 75}


### 2.3 Basic File Handling

File handling is essential in Data Science for reading dataset files (CSVs, TXT, JSON) and writing logs or processed results. Using Python's context manager (`with` statement) ensures files are properly opened and closed automatically.

In [ ]:
# 1. File Write Karna (data/ folder ke andar file banayein)
file_path = "../data/sample_log.txt"

with open(file_path, "w") as f:
    f.write("YuvaIntern AI/ML Task 1 - Data Science Fundamentals\n")
    f.write("File handling completed successfully.")

# 2. File Read Karna
with open(file_path, "r") as f:
    content = f.read()

print("File Content:\n", content)

### 2.3 Basic File Handling

File handling is essential in Data Science for reading dataset files (CSVs, TXT, JSON) and writing logs or processed results. Using Python's context manager (`with` statement) ensures files are properly opened and closed automatically.

In [10]:
# File path setup
file_path = "../data/sample_log.txt"

# 1. Writing to a file
with open(file_path, "w") as f:
    f.write("YuvaIntern AI/ML Task 1 - Data Science Fundamentals\n")
    f.write("File handling completed successfully.")

# 2. Reading from the file
with open(file_path, "r") as f:
    content = f.read()

print("--- File Content Output ---")
print(content)

--- File Content Output ---
YuvaIntern AI/ML Task 1 - Data Science Fundamentals
File handling completed successfully.


## 3. Statistical Foundations

Statistics is the backbone of Data Science. It provides the methods to summarize data, identify trends, measure variability, and draw meaningful conclusions from datasets.

### 3.1 Measures of Central Tendency & Dispersion
- **Mean:** The arithmetic average of a dataset.
- **Median:** The middle value in a sorted dataset (robust to outliers).
- **Mode:** The most frequently occurring value(s).
- **Variance:** Measures how far set numbers are spread out from their average.
- **Standard Deviation:** Square root of variance; represents dispersion in the same unit as the original data.
- **Percentiles:** Indicates the value below which a given percentage of observations fall (e.g., 25th, 50th, 75th percentiles).

In [12]:
import numpy as np

# Dataset
scores = [75, 80, 85, 90, 85, 95, 70, 85, 100, 60]

# Descriptive Statistics
mean_val = np.mean(scores)
median_val = np.median(scores)
mode_val = max(set(scores), key=scores.count)
var_val = np.var(scores)
std_val = np.std(scores)
p25 = np.percentile(scores, 25)
p75 = np.percentile(scores, 75)

print("--- Descriptive Statistics ---")
print("Mean:", mean_val)
print("Median:", median_val)
print("Mode:", mode_val)
print("Variance:", var_val)
print("Standard Deviation:", std_val)
print("25th Percentile:", p25)
print("75th Percentile:", p75)

--- Descriptive Statistics ---
Mean: 82.5
Median: 85.0
Mode: 85
Variance: 126.25
Standard Deviation: 11.236102527122116
25th Percentile: 76.25
75th Percentile: 88.75


## 3. Statistical Foundations

Statistics is the backbone of Data Science. It provides the methods to summarize data, identify trends, measure variability, and draw meaningful conclusions from datasets.

### 3.1 Measures of Central Tendency & Dispersion
- **Mean:** The arithmetic average of a dataset.
- **Median:** The middle value in a sorted dataset (robust to outliers).
- **Mode:** The most frequently occurring value(s).
- **Variance:** Measures how far numbers are spread out from their average.
- **Standard Deviation:** Square root of variance; measures dispersion in same units as raw data.
- **Percentiles:** Indicates the value below which a given percentage of observations fall.

In [14]:
import numpy as np

# Dataset: Student scores
scores = [75, 80, 85, 90, 85, 95, 70, 85, 100, 60]

# Calculate Statistical Measures
mean_val = np.mean(scores)
median_val = np.median(scores)
mode_val = max(set(scores), key=scores.count)
var_val = np.var(scores)
std_val = np.std(scores)
p25 = np.percentile(scores, 25)
p75 = np.percentile(scores, 75)

print("--- Descriptive Statistics Output ---")
print(f"Mean: {mean_val}")
print(f"Median: {median_val}")
print(f"Mode: {mode_val}")
print(f"Variance: {var_val:.2f}")
print(f"Standard Deviation: {std_val:.2f}")
print(f"25th Percentile: {p25}")
print(f"75th Percentile: {p75}")

--- Descriptive Statistics Output ---
Mean: 82.5
Median: 85.0
Mode: 85
Variance: 126.25
Standard Deviation: 11.24
25th Percentile: 76.25
75th Percentile: 88.75


### 3.2 Probability & Probability Distributions

Probability measures the likelihood of an event occurring. In Data Science, understanding data distributions helps in making inferences and selecting appropriate machine learning algorithms.

- **Normal Distribution (Gaussian):** Bell-shaped curve symmetric around the mean (Mean = Median = Mode). Standard deviation controls spread ($68\%-95\%-99.7\%$ rule).
- **Binomial Distribution:** Models the number of successes in a fixed number of independent binary (Yes/No) trials.
- **Poisson Distribution:** Models the count of events occurring within a fixed interval of time or space.

In [15]:
import numpy as np

# Reproducibility ke liye seed set karein
np.random.seed(42)

# 1. Normal Distribution (loc=mean, scale=std, size=samples)
normal_data = np.random.normal(loc=100, scale=15, size=1000)

# 2. Binomial Distribution (n=trials, p=probability, size=samples)
binomial_data = np.random.binomial(n=10, p=0.5, size=1000)

# 3. Poisson Distribution (lam=average rate, size=samples)
poisson_data = np.random.poisson(lam=5, size=1000)

print("--- Probability Distributions Summary ---")
print(f"Normal Dist -> Mean: {np.mean(normal_data):.2f}, Std Dev: {np.std(normal_data):.2f}")
print(f"Binomial Dist -> Mean: {np.mean(binomial_data):.2f}, Std Dev: {np.std(binomial_data):.2f}")
print(f"Poisson Dist -> Mean: {np.mean(poisson_data):.2f}, Std Dev: {np.std(poisson_data):.2f}")

--- Probability Distributions Summary ---
Normal Dist -> Mean: 100.29, Std Dev: 14.68
Binomial Dist -> Mean: 5.03, Std Dev: 1.57
Poisson Dist -> Mean: 4.93, Std Dev: 2.10


### 3.3 Hypothesis Testing, p-values & Correlation

- **Hypothesis Testing:** A statistical method to determine if there is enough evidence to reject a null hypothesis ($H_0$).
- **p-value:** The probability of obtaining results at least as extreme as the observed results, assuming $H_0$ is true. A $p$-value $< 0.05$ usually indicates statistical significance.
- **Correlation:** Measures the strength and direction of a linear relationship between two variables (ranges from $-1$ to $+1$).

In [19]:
import numpy as np
from math import erf, sqrt


np.random.seed(42)

# Sample Data Generation
group_a = np.random.normal(loc=70, scale=10, size=50) # Score before training
group_b = np.random.normal(loc=78, scale=10, size=50) # Score after training

# 1. Hypothesis Testing (independent samples t-test)
mean_difference = np.mean(group_a) - np.mean(group_b)
pooled_variance = (
    ((len(group_a) - 1) * np.var(group_a, ddof=1))
    + ((len(group_b) - 1) * np.var(group_b, ddof=1))
) / (len(group_a) + len(group_b) - 2)
standard_error = sqrt(pooled_variance * (1 / len(group_a) + 1 / len(group_b)))
t_stat = mean_difference / standard_error
normal_cdf = lambda value: 0.5 * (1 + erf(value / sqrt(2)))
p_value = 2 * (1 - normal_cdf(abs(t_stat)))

# 2. Correlation Matrix
x = np.array([10, 20, 30, 40, 50, 60])
y = np.array([15, 25, 32, 45, 51, 65])
corr_matrix = np.corrcoef(x, y)
correlation = corr_matrix[0, 1]

print("--- Statistical Tests Output ---")
print(f"t-statistic: {t_stat:.4f}")
print(f"p-value: {p_value:.4f}")
if p_value < 0.05:
    print("Result: Reject Null Hypothesis (Significant difference between groups)")
else:
    print("Result: Fail to Reject Null Hypothesis (No significant difference)")

print(f"\nPearson Correlation: {correlation:.4f}")

--- Statistical Tests Output ---
t-statistic: -5.7671
p-value: 0.0000
Result: Reject Null Hypothesis (Significant difference between groups)

Pearson Correlation: 0.9953


## 4. Data Manipulation with Pandas

Pandas is the primary library for data wrangling, cleaning, and analysis in Python. It introduces two core data structures:
- **Series:** 1D labeled array capable of holding any data type.
- **DataFrame:** 2D tabular data structure with labeled axes (rows and columns).

### 4.1 Series & DataFrames Creation and Inspection
- Creating DataFrames from Dictionaries/CSVs
- Inspecting data shape, column data types, `.head()`, `.tail()`, `.info()`, and `.describe()`

In [20]:
import pandas as pd

# 1. Dictionary se DataFrame banayein
data = {
    "StudentID": [101, 102, 103, 104, 105],
    "Name": ["Amit", "Neha", "Pooja", "Rohan", "Suresh"],
    "Department": ["CS", "ECE", "CS", "ME", "ECE"],
    "Score": [85, 92, 78, 65, 88],
    "Passed": [True, True, True, False, True]
}

df = pd.DataFrame(data)

# 2. DataFrame Inspection
print("--- First 3 Rows (.head) ---")
print(df.head(3))

print("\n--- DataFrame Summary (.info) ---")
df.info()

print("\n--- Statistical Summary (.describe) ---")
print(df.describe())

--- First 3 Rows (.head) ---
   StudentID   Name Department  Score  Passed
0        101   Amit         CS     85    True
1        102   Neha        ECE     92    True
2        103  Pooja         CS     78    True

--- DataFrame Summary (.info) ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   StudentID   5 non-null      int64 
 1   Name        5 non-null      object
 2   Department  5 non-null      object
 3   Score       5 non-null      int64 
 4   Passed      5 non-null      bool  
dtypes: bool(1), int64(2), object(2)
memory usage: 297.0+ bytes

--- Statistical Summary (.describe) ---
        StudentID      Score
count    5.000000   5.000000
mean   103.000000  81.600000
std      1.581139  10.597169
min    101.000000  65.000000
25%    102.000000  78.000000
50%    103.000000  85.000000
75%    104.000000  88.000000
max    105.000000  92.000000


### 4.2 Indexing, Selection, and Filtering

Pandas provides powerful tools to select specific subsets of data:
- **Column Selection:** `df['ColumnName']` or `df[['Col1', 'Col2']]`
- **Label-based Indexing (`.loc`):** Access rows and columns by explicit labels/names.
- **Position-based Indexing (`.iloc`):** Access rows and columns by integer positions (0-indexed).
- **Boolean Masking:** Filter rows based on conditional logic (e.g., `df[df['Score'] > 80]`).

In [21]:
# Assuming 'df' is available from the previous cell

# 1. Column Selection
scores_series = df['Score']
subset_df = df[['Name', 'Department', 'Score']]

# 2. .loc (Label-based) & .iloc (Position-based)
loc_subset = df.loc[0:2, ['Name', 'Score']]    # Rows index 0 to 2, specific columns
iloc_subset = df.iloc[0:3, 1:4]               # Rows 0 to 2, Columns index 1 to 3

# 3. Filtering with Boolean Indexing
high_scorers = df[df['Score'] >= 80]
cs_students = df[(df['Department'] == 'CS') & (df['Score'] > 75)]

print("--- Filtered: Score >= 80 ---")
print(high_scorers)

print("\n--- Filtered: CS Dept & Score > 75 ---")
print(cs_students)

--- Filtered: Score >= 80 ---
   StudentID    Name Department  Score  Passed
0        101    Amit         CS     85    True
1        102    Neha        ECE     92    True
4        105  Suresh        ECE     88    True

--- Filtered: CS Dept & Score > 75 ---
   StudentID   Name Department  Score  Passed
0        101   Amit         CS     85    True
2        103  Pooja         CS     78    True


### 4.3 Grouping, Aggregation & Data Cleaning

Data aggregation allows us to summarize metrics across categories, while data cleaning handles missing or duplicate entries.

- **Grouping (`.groupby()`):** Splits data into groups based on criteria, applies a function (mean, sum, count), and combines results.
- **Missing Values:** Identified using `.isnull()`, handled via `.dropna()` (removal) or `.fillna()` (imputation).
- **Sorting & Renaming:** Reordering rows with `.sort_values()` and modifying column names with `.rename()`.

In [22]:
import pandas as pd
import numpy as np

# 1. GroupBy Aggregation (Department-wise mean & count)
dept_summary = df.groupby('Department')['Score'].agg(['mean', 'count', 'max']).reset_index()

# 2. Introducing & Handling Missing Data
df_missing = df.copy()
df_missing.loc[1, 'Score'] = np.nan  # Inject NaN in row 1

# Missing values check
null_counts = df_missing.isnull().sum()

# Imputation (Filling missing score with mean score)
mean_score = df_missing['Score'].mean()
df_filled = df_missing.copy()
df_filled['Score'] = df_filled['Score'].fillna(mean_score)

print("--- Department Group Summary ---")
print(dept_summary)

print("\n--- Missing Value Count ---")
print(null_counts)

print("\n--- DataFrame After Mean Imputation ---")
print(df_filled)

--- Department Group Summary ---
  Department  mean  count  max
0         CS  81.5      2   85
1        ECE  90.0      2   92
2         ME  65.0      1   65

--- Missing Value Count ---
StudentID     0
Name          0
Department    0
Score         1
Passed        0
dtype: int64

--- DataFrame After Mean Imputation ---
   StudentID    Name Department  Score  Passed
0        101    Amit         CS   85.0    True
1        102    Neha        ECE   79.0    True
2        103   Pooja         CS   78.0    True
3        104   Rohan         ME   65.0   False
4        105  Suresh        ECE   88.0    True


## 5. Data Visualization

Data visualization transforms numerical results into intuitive visual patterns, aiding exploratory data analysis (EDA) and storytelling.

### 5.1 Exploratory Visualizations
- **Line Plot:** Continuous trends over time or ordered sequences.
- **Bar Plot:** Categorical variable comparisons.
- **Histogram & KDE:** Distribution shape, skewness, and frequency.
- **Scatter Plot:** Bivariate relationship and correlation between numerical variables.

import matplotlib.pyplot as plt
import seaborn as sns

# Plot style set karein
sns.set_theme(style="whitegrid")
fig, axes = plt.subplots(2, 2, figsize=(12, 10))

# 1. Bar Plot (Department vs Mean Score)
sns.barplot(data=df, x="Department", y="Score", ax=axes[0, 0], palette="viridis", errorbar=None)
axes[0, 0].set_title("Average Score by Department")

# 2. Histogram with KDE (Distribution of Scores)
sns.histplot(df["Score"], kde=True, ax=axes[0, 1], color="teal")
axes[0, 1].set_title("Score Distribution")

# 3. Scatter Plot (StudentID vs Score)
sns.scatterplot(data=df, x="StudentID", y="Score", hue="Department", s=100, ax=axes[1, 0])
axes[1, 0].set_title("Student ID vs Score")

# 4. Line Plot (Trend)
axes[1, 1].plot(df["Name"], df["Score"], marker="o", color="crimson", linewidth=2)
axes[1, 1].set_title("Score Trend by Student")

plt.tight_layout()
plt.show()
print("Plots generated successfully.")

In [23]:
# Assuming 'df' is available from the previous cell

# 1. Column Selection
scores_series = df['Score']
subset_df = df[['Name', 'Department', 'Score']]

# 2. .loc (Label-based) & .iloc (Position-based)
loc_subset = df.loc[0:2, ['Name', 'Score']]    # Rows index 0 to 2, specific columns
iloc_subset = df.iloc[0:3, 1:4]               # Rows 0 to 2, Columns index 1 to 3

# 3. Filtering with Boolean Indexing
high_scorers = df[df['Score'] >= 80]
cs_students = df[(df['Department'] == 'CS') & (df['Score'] > 75)]

print("--- Filtered: Score >= 80 ---")
print(high_scorers)

print("\n--- Filtered: CS Dept & Score > 75 ---")
print(cs_students)

--- Filtered: Score >= 80 ---
   StudentID    Name Department  Score  Passed
0        101    Amit         CS     85    True
1        102    Neha        ECE     92    True
4        105  Suresh        ECE     88    True

--- Filtered: CS Dept & Score > 75 ---
   StudentID   Name Department  Score  Passed
0        101   Amit         CS     85    True
2        103  Pooja         CS     78    True


### 4.3 Grouping, Aggregation & Data Cleaning

Data aggregation allows us to summarize metrics across categories, while data cleaning handles missing or duplicate entries.

- **Grouping (`.groupby()`):** Splits data into groups based on criteria, applies a function (mean, sum, count), and combines results.
- **Missing Values:** Identified using `.isnull()`, handled via `.dropna()` (removal) or `.fillna()` (imputation).

In [24]:
import pandas as pd
import numpy as np

# 1. GroupBy Aggregation (Department-wise mean & count)
dept_summary = df.groupby('Department')['Score'].agg(['mean', 'count', 'max']).reset_index()

# 2. Introducing & Handling Missing Data
df_missing = df.copy()
df_missing.loc[1, 'Score'] = np.nan  # Inject NaN in row 1

# Missing values check
null_counts = df_missing.isnull().sum()

# Imputation (Filling missing score with mean score)
mean_score = df_missing['Score'].mean()
df_filled = df_missing.copy()
df_filled['Score'] = df_filled['Score'].fillna(mean_score)

print("--- Department Group Summary ---")
print(dept_summary)

print("\n--- Missing Value Count ---")
print(null_counts)

print("\n--- DataFrame After Mean Imputation ---")
print(df_filled)

--- Department Group Summary ---
  Department  mean  count  max
0         CS  81.5      2   85
1        ECE  90.0      2   92
2         ME  65.0      1   65

--- Missing Value Count ---
StudentID     0
Name          0
Department    0
Score         1
Passed        0
dtype: int64

--- DataFrame After Mean Imputation ---
   StudentID    Name Department  Score  Passed
0        101    Amit         CS   85.0    True
1        102    Neha        ECE   79.0    True
2        103   Pooja         CS   78.0    True
3        104   Rohan         ME   65.0   False
4        105  Suresh        ECE   88.0    True


## 5. Data Visualization

Data visualization transforms numerical results into intuitive visual patterns, aiding exploratory data analysis (EDA) and storytelling.

### 5.1 Exploratory Visualizations
- **Line Plot:** Continuous trends over time or ordered sequences.
- **Bar Plot:** Categorical variable comparisons.
- **Histogram & KDE:** Distribution shape, skewness, and frequency.
- **Scatter Plot:** Bivariate relationship and correlation between numerical variables.

In [26]:
# Assuming 'df' is available from the previous cell

print("--- Average Score by Department ---")
print(df.groupby("Department")["Score"].mean())

print("\n--- Score Distribution ---")
score_bins = np.histogram(df["Score"], bins=5)[0]
for bin_number, frequency in enumerate(score_bins, start=1):
    print(f"Bin {bin_number}: {'#' * frequency} ({frequency})")

print("\n--- Student ID vs Score ---")
print(df[["StudentID", "Department", "Score"]].to_string(index=False))

print("\n--- Score Trend by Student ---")
for name, score in zip(df["Name"], df["Score"]):
    print(f"{name:<10} {'#' * (score // 5)} {score}")

print("\nVisualization summary generated successfully.")

--- Average Score by Department ---
Department
CS     81.5
ECE    90.0
ME     65.0
Name: Score, dtype: float64

--- Score Distribution ---
Bin 1: # (1)
Bin 2:  (0)
Bin 3: # (1)
Bin 4: # (1)
Bin 5: ## (2)

--- Student ID vs Score ---
 StudentID Department  Score
       101         CS     85
       102        ECE     92
       103         CS     78
       104         ME     65
       105        ECE     88

--- Score Trend by Student ---
Amit       ################# 85
Neha       ################## 92
Pooja      ############### 78
Rohan      ############# 65
Suresh     ################# 88

Visualization summary generated successfully.


## 6. Summary & Key Findings

### Key Takeaways:
1. **Python Fundamentals:** Standardized coding conventions, dynamic data structures, and safe file handling (`with` statement) established a clean code foundation.
2. **Statistical Foundations:** Descriptive statistics confirmed a central tendency of ~82.5 for student test scores. T-tests demonstrated statistically significant variations across groups ($p < 0.05$).
3. **Data Wrangling:** Pandas operations (`.loc`, `.iloc`, `.groupby()`) enabled structured manipulation, and missing values were dynamically imputed using mean strategies without data loss.
4. **Data Visualization:** Matplotlib and Seaborn successfully highlighted performance trends across departments, showing higher score density in the ECE stream.